### ISSUE TITLE
I can't edit a breakpoint expression using orca.

### ISSUE THREAD

**Issue by jvesouza** [microsoft/vscode#106505]
Issue Type: <b>Bug</b>

CC: @isidorn @joanmarie

1. Launch orca.
1. Launch vscode.
1. Create a .js file.
1. Insert the following code:
`<a = 1>`

1. Activate the run menu.
1. Expand the new breakpoint menu.
1. Choose the log point option.
1. Type for example "this is a test".
1. Try to review the typed text using left and right arrow.

In my environment nothing is read by orca.

VS Code version: Code - Insiders 1.50.0-insider (bc38b0accb3c62180317c63dbe7323200977db0f, 2020-09-11T12:27:36.137Z)
OS version: Linux x64 5.8.7-arch1-1

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz (4 x 2494)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: disabled_software<br>skia_renderer: enabled_on<br>video_decode: unavailable_off<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 1, 1|
|Memory (System)|7.68GB (0.46GB free)|
|Process Argv|. --no-sandbox --crash-reporter-id bcecf5e0-97d1-4830-b28c-0943c4adab1c --crash-reporter-id bcecf5e0-97d1-4830-b28c-0943c4adab1c|
|Screen Reader|yes|
|VM|0%|
|DESKTOP_SESSION|gnome-xorg|
|XDG_CURRENT_DESKTOP|GNOME|
|XDG_SESSION_DESKTOP|gnome-xorg|
|XDG_SESSION_TYPE|x11|
</details>

**Comment by isidorn** [comment]
@jvesouza thanks for filling this.
I can reproduce this. However I can only reproduce with Orca, not with VoiceOver or NVDA.
@joanmarie the issue here might be that we change focus to the "breakpoint editor" with a timeout and for some weird reason that might confuse orca. Other than that there is nothing special here, the "breakpoint editor" is the same editor as the regular one for which Orca works just fine. Can you maybe take a look on your side?

**Comment by joanmarie** [comment]
I can confirm the problem. Chromium (or VSCode) is not emitting caret-moved events as I arrow within the breakpoint expression. Thus Orca has no idea the caret moved or that something should be presented.

**Comment by isidorn** [comment]
@joanmarie thanks for taking a look.
Weirdly this works just fine on macOS with VoiceOver, and with NVDA things get read, but when you navigate left/right always the first character gets read.
Will investigate more...

**Comment by joanmarie** [comment]
@isidorn I'll see if I can dig into it from the Chromium side of things. Working on a different Chromium bug atm which I need to finish first, however. If you don't make progress and I space it out, please ping me.

**Comment by joanmarie** [comment]
@isidorn Sorry for the delay. The bug I had been working on turned out to be rather rabbit holey. Now that I'm looking into this I have a question: How can I reproduce this problem in the browser (i.e. via `yarn web`)? In the Run menu, the only option I see is "Install Additional Debuggers..." which brings me to the completely-empty Extensions Marketplace.

**Comment by isidorn** [comment]
@joanmarie sorry for the delay, I was on vacation...
Good question. This is the simplest way I could make it work:
1) Make the following change [here](https://github.com/microsoft/vscode/blob/49d9c8c79b6fd9aaeeb5e869e37d93cfd0e7b703/src/vs/workbench/contrib/debug/browser/debugConfigurationManager.ts#L117) change `>` to `>=`
2) in `yarn web` vscode instance have the following setting `"debug.allowBreakpointsEverywhere": true`

Now you can set breakpoint in any file and should be able to reproduce the issue.
Sorry that there is no simpler way.

**Comment by joanmarie** [comment]
@isidorn Thanks! I made those changes, but how do I set the breakpoint? According to the steps to reproduce, it's in the run menu. But the only thing I have in the run menu is "Install Additional Debuggers..."

**Comment by joanmarie** [comment]
Oops, clicking to the left of the line number does it. Sorry for the noise!

**Comment by joanmarie** [comment]
Ok, so some updates:
* Good: In the latest Chromium, I am seeing caret-moved events being emitted. So there may be a change, at least for Linux, which we can backport. In fact, I think it's a change I made. But looking into this bug here, we have a use case I don't like and might wish to rethink that change. ;) In the meantime....
* Good: Caret-moved events are initially coming from the textarea, BUT:
* Bad(ish): That textarea is hidden so it cannot emit events itself.
* Bad(ish): There are other things which cannot emit caret-moved events itself, such as the inline text box descendant of a text node. So the ATK support looks for the first parent which can emit caret-moved events. Ascending the accessibility tree in those non-hidden cases seems to work as expected. But in the case of the accessible-but-hidden textarea, the first accessible ancestor which passes the can-emit-caret-moved-events test happens to be a div with class of overlayWidgets. Yikes! This is the thing I might wish to rethink, btw.
* When Orca gets the caret-moved event, it tries to figure out where it's really coming from, and it does conclude it's the textarea. But because that textarea is hidden, it cannot have the caret and so it accepts that the caret has moved in that ancestor div and says nothing. I *could* change that potentially, but it seems like a hack(?).
* We can potentially fix this in Chromium itself by allowing *certain* hidden elements to emit text events. And I guess I'll see about doing that and then making a case for it to Google. If they don't have a problem with hidden inputs emitting events, then I guess I don't either. Though admittedly right now I'm not super jazzed by it. :wink:

**Comment by isidorn** [comment]
Thanks @joanmarie for looking more into this and for the write-up.

Motivation for allowing hidden elements to emit text events: there might be use cases like vscode that use a rich editor on top, but have hidden textArea fields just for accessibility. Not sure if anyone else is currently doing this pattern, but I can imagine others doing the same.

What I do not understand is why this works for our regular editor but does not work for the breakpoint edit box? What is different between those two? Since on our side we use the same editor widget in both cases that use a hidden `textArea` element.

If there is a change to be backported from Chrome let me know and we can try to do it no problem.

Thanks!

**Comment by joanmarie** [comment]
@isidorn There is something causing the regular editor to lack the "hidden" property, whereas the breakpoint edit box has the "hidden" property. Will see if I can figure out what is causing this difference.

**Comment by isidorn** [comment]
Ok. From our side they should be the same as they have same html attributes set hopefully.

**Comment by joanmarie** [comment]
@isidorn 

So the breakpoint editor appears to descend from this element:

```
<div aria-hidden="true" role="presentation" class="zone-widget" widgetid="vs.editor.contrib.zoneWidgeth1" style="width: 456px; left: 0px; top: 19px; height: 38px; position: absolute;">
```

And according to the ARIA spec:

> An element is considered hidden if it, or any of its ancestors are not rendered or have their aria-hidden attribute value set to true.

Why is `aria-hidden` on that ancestor div of the breakpoint editor?

I'm not seeing `aria-hidden` on any ancestor of the code editor. So I *think* that's the difference.

**Comment by isidorn** [comment]
@joanmarie awesome catch, somehow I missed that. My brain is still in vacation mode... I was not aware that some of our editor zone widgets have this flag to be "hidden from screen readers" becasue there are alternative ways to access their data.

Anyways, thanks to your smartness I pushed a fix for this. Now it should work I believe, you can verify and let me know.

**Comment by joanmarie** [comment]
Works for me. Thanks!

**Comment by isidorn** [comment]
Awesome, adding verified label.
@jvesouza you can try it out in vscode insiders from tomorrow and if it still does not work let us know.

**Comment by jvesouza** [comment]
It looks great.
@joanmarie @isidorn Thanks, great job.
